gcsfs/core.py

```diff
@@ -398,65 +562,77 @@ def rmdir(self, bucket):
-        """
-        if bucket in {'/', '', None}:
-            self.dirs.clear()
-        else:
-            self.dirs.pop(bucket, None)
+        self.invalidate_cache(bucket)
 
     def ls(self, path, detail=False):
-        if path in ['', '/']:
-            out = self._list_buckets()
-        else:
-            bucket, prefix = split_path(path)
-            path = '/'.join([bucket, prefix])
-            files = self._list_bucket(bucket)
-            seek, l, bit = (path, len(path), '') if path.endswith('/') else (
-                path+'/', len(path)+1, '/')
-            out = []
-            for f in files:
-                if (f['name'].startswith(seek) and '/' not in f['name'][l:] or
-                        f['name'] == path):
-                    out.append(f)
-                elif f['name'].startswith(seek) and '/' in f['name'][l:]:
-                    directory = {
-                        'bucket': bucket, 'kind': 'storage#object',
-                        'size': 0, 'storageClass': 'DIRECTORY',
-                        'name': path+bit+f['name'][l:].split('/', 1)[0]+'/'}
-                    if directory not in out:
-                        out.append(directory)
-        if detail:
-            return out
+        """List objects under the given '/{bucket}/{prefix} path."""
+        if path in ['/', '']:
+            out = self.buckets
         else:
-            return [f['name'] for f in out]
+            if not path.endswith("/"):
+                path = path + "/"
+
+            listing = self._list_objects(path)
+            bucket, key = split_path(path)
+
+            if not detail:
+                result = []
+
+                # Convert item listing into list of 'item' and 'subdir/'
+                # entries. Items may be of form "key/", in which case there
+                # will be duplicate entries in prefix and item_names.
+                item_names = [
+                    f["name"][len(key):] for f in listing["items"]
+                    if f["name"][len(key):]
+                ]
+                prefixes = [p for p in listing["prefixes"]]
+
+                return list(set(item_names + prefixes))
+
+            else:
+                item_details = listing["items"]
+                pseudodirs = [{
+                        'bucket': bucket,
+                        'name': key + prefix,
+                        'kind': 'storage#object',
+                        'size': 0,
+                        'storageClass': 'DIRECTORY',
+                    }
+                    for prefix in listing["prefixes"]
+                ]
+
+                return item_details + pseudodirs
 
     def walk(self, path, detail=False):
+        """ Return all real keys belows path. """
         bucket, prefix = split_path(path)
+
         if not bucket:
-            raise ValueError('Cannot walk all of GCS')
+            raise ValueError(
+                "walk path must include target bucket: %s" % path)
```

Comment:
`path` is always empty here, so it is not very useful to report it. "Path must include at least a bucket" ?